Storage model update

jwangdx opened the pull request:
### Description

Made several changes to the storage model. For detailed information, please read [this doc](http://md.dxchain.com/yz5reGBNThqOqmWaYj5OCg?view)

#### New features

1. Some user-defined variables changed to constants for user-friendliness
2. When user `SetRentPayment`, the expectedStorage, expectedUploads e.t.c are automatically estimated.
3. Added an prompt after user set host config in console to remind the user to execute `shost.announce()`
4. Added the functionality of `HostManager` in storage host manager to estimate the market prices.

#### Modification

1. Host and client default params are updated.
2. The logic for storage client period calculation is updated.
3. Updated the logic for renewing the contract when the contract fund is not enough.
4. Updated the criteria for client removing the host from `storageHostManager`
5. Refactored part of host evaluation and scan logics.

#### Bug fixes

1. Fix an error in `waitScanFinish`. The finishing condition should add `numWorkers == 0`

### Type of change

- [x] New feature
- [x] Bug fix

### Testing

Implemented multiple test cases:

* TestStorageHostManager_scanLogic (`storage/storageclient/storagehostmanager/scan_test.go:53`)
* TestStorageHostManager_isInitialScanFinished (`storage/storageclient/storagehostmanager/storagehostmanager_test.go:134`)
* TestStorageHostManager_finishInitialScan (storage/storageclient/storagehostmanager/storagehostmanager_test.go:155)
* storage/storageclient/storagehostmanager/hostmarket_test.go

xzhangdx commented on storage/storageclient/contractmanager/contractmanager_test.go:
```diff
@@ -10,6 +10,8 @@ import (
 	"testing"
 
+	"github.com/DxChainNetwork/godx/common/unit"
+
```
go import format: remove the line between internal packages

jwangdx replied:
Agreed and updated

xzhangdx commented on storage/storageclient/storagehostmanager/hostmarket_test.go:
```diff
@@ -5,7 +5,16 @@
+	"time"
+
+	"github.com/DxChainNetwork/godx/p2p/enode"
+
```
go import format: remove the line between internal packages

jwangdx replied:
Agreed and updated

xzhangdx commented on storage/storageclient/storagehostmanager/scan_test.go:
```diff
@@ -6,10 +6,15 @@ package storagehostmanager
 
+	"github.com/Pallinder/go-randomdata"
+
 	"github.com/DxChainNetwork/godx/accounts"
```
go import format: the third packages should be under the internal packages

jwangdx replied:
Updated

xzhangdx commented on storage/storageclient/storagehosttree/storagehosttree.go:
```diff
@@ -112,15 +109,15 @@ func (t *StorageHostTree) Remove(enodeID enode.ID) error {
 
 // all will retrieve, sort, and return all host information stored in the tree
-func (t *StorageHostTree) all() (his []storage.HostInfo) {
+func (t *storageHostTree) all() (his []storage.HostInfo) {
```
why some funcs are also public and prive at the same time? If already public, why not integrate them to public one?

jwangdx replied:
The private function `all` comes from an historic reason. 

The private function `all` was used to be used in a private function of storageHostTree named `setEvaluator`, which is removed in the latest update. This function needs to get the all entries without lock. 

Now private function `all` is only used in `All`, thus it would be reasonable to remove the private function. Thanks for advice.

xfliudx commented on storage/types.go:
```diff
@@ -410,6 +409,18 @@ type (
 	}
 )
 
+type (
```
this type style mismatch its following. suggest uniform style or union type struct as one

jwangdx replied:
Updated. Moved the structure to union.

jwangdx commented:
> 1. `storage/types.go:212` RentPaymentAPIDisplay data structure needs to be changed as well since you modified the RentPayment data structure (remove the RenewWindow field).
> 2. `storage/storageclient/contractmanager/rentpayment.go:128` suggest to modify this error message since `RenewWindow` is not allowed to be configured. For instance, storage period must be greater than 12 hours.
> 3. `storage/storageclient/clientsettingparse.go:111` remove unused method
> 4. `cmd/gdx/sclientcmd.go:47` remove the renew flag
> 5. `cmd/gdx/sclientcmd.go:643` remove the renew flag checking
> 6. `cmd/gdx/sclientcmd.go:223` update the setConfig description
> 7. `cmd/gdx/sclientcmd.go:219` remove the renew flag here as well

Agreed and updated.

jwangdx commented:
> 1. `storage/storageclient/storagehosttree/storagehosttree.go:140` what is the point of returning the evaluation in the `RetrieveHostInfo` method. Based on my observation, the only time you used the evaluation is in the test case. Suggest to implement another method to explicitly return the evaluation.
> 2. `storage/storageclient/storagehostmanager/defaults.go:25` is there a reason to change the `scanCheckDuration` from 1 second to 200 millisecond.
> 3. `storage/storageclient/storagehostmanager/hostmarket.go:136` why return 0? If 0 is returned, then each field in the MarketPrice can potentially be 0.
> 4. `storage/storageclient/storagehostmanager/hostmarket.go:147` why don't you just write one sort method and pass in the variable that needs to be compared? Instead, you implemented 6 similar sorting methods.

1. The test case use the evaluation in the upper function call. If you are to implement a new public function to return the evaluation of a host, you need to add the signature of the new function to interface StorageHostTree which is only used in test case. Comparing adding a new return value with adding a new function, I choose the former one.
2. Yes. The time for running the test cases in storageHostManager module will decrease from 3s to 1.6s.
3. Theoretically, the code in this line should never be executed since the only condition for highIndex == lowIndex is that the size of the hostInfos is 0, which is already handled in previous code. But you are right, It would be better to return the default value than 0. Will fix in later commits.
4.  Because of the performance. If I am not understanding wrong, what you want is to implement a sorting method for list of common.BigInt. When estimating the average price, make a new list of common.BigInt and then get the average from the list. This could work. But what you have done here is copying the content of the list again and again. In my method, instead of copying the content, only the pointers are copied and compared. Moreover, you have to write six functions each with a for loop for copying the expected fields to a list []common.BigInt, each function with a for loop.

jwangdx commented:
> 1. `storage/storageclient/contractmanager/rentpayment.go:116` for `RentPaymentValidation`, there is no need to return error when ExpectedStorage, ExpectedUpload, and etc. are 0 value because the user is not able to configure those values. Instead, we should assign those values to default values.
> 2. `storage/storageclient/contractmanager/contractcreate.go:34` the endHeight needs to be changed since the method for calculating the currentPeriod is changed.
> 3. `storage/storagehost/api.go:259` storage host window size is no longer configurable by the user.

1. You are right, removed the validations for these fields.
2. The current end height calculation is correct. Please recheck the period calculation logic.
3. Removed the used function.

xfliudx commented:
storage/storageclient/storagehostmanager/hostmarket.go:28
计算MarketPrice时，可以单独开启一个goroutine，定时更新markerPrice；在获取时直接获取其数据即可，这样就可以避免calculateMarketPrice的阻塞

其他逻辑我这边没什么问题了

xfliudx commented:
1. Storage/storageclient/storagehostmanager/hostmarket.go:28
When calculating MarketPrice, you can open a goroutine separately, update the markerPrice regularly; you can get its data directly when you get it, so you can avoid blocking of calculateMarketPrice.

I have no problem about other logic code

jwangdx commented:
> 1. Storage/storageclient/storagehostmanager/hostmarket.go:28
>    When calculating MarketPrice, you can open a goroutine separately, update the markerPrice regularly; you can get its data directly when you get it, so you can avoid blocking of calculateMarketPrice.
> 
> I have no problem about other logic code

I quite agree with this design. Although currently the price is stored in cached field, it would be better to be update within a goroutine. Thanks for advise.

jwangdx commented on storage/storageclient/contractmanager/rentpayment.go:
```diff
@@ -7,17 +7,36 @@ package contractmanager
 	"reflect"
+
+	"github.com/DxChainNetwork/godx/common/unit"
+
```
Updated

jwangdx commented:
> 1. `storage/storagehost/defaults.go:45` remove the unused constant `proofWindowSize`
> 2. `storage/storageclient/storagehostmanager/hostmarket.go:80` suggest to change all methods to `getAveragePriceByField` as well.
> 
> :shipit:

Updated.
